### PR TITLE
[AutoParallel] support gqa for fused_rope and flash_attention spmd rules

### PR DESCRIPTION
**Pull request by deepllz**
### PR types

 Bug fixes 
### PR changes

Others
### Description

Pcard-73145
fused_rope和flash_attention的切分推导规则支持GQA

### CODE REVIEW

**Review by liuzhenhai93 — APPROVED** [review]
LGTM

File: paddle/phi/infermeta/spmd_rules/fused_rope.cc [hunk 1]
```diff
@@ -68,13 +68,35 @@ void check_k_or_v(const DistMetaTensor& k_or_v,
...
   PADDLE_ENFORCE_EQ(
-      shape,
-      q_shape,
+      q_num_head % k_num_head == 0,
+      true,
       phi::errors::InvalidArgument(
-          "The shape of q and k/v's are not matched, [%d]  vs [%d]",
-          str_join(q_shape),
-          str_join(shape)));
+          "The num_head of q must be divisible by k/v's, but [%d]  vs [%d]",
```
Comment (zhiqiu) [inline paddle/phi/infermeta/spmd_rules/fused_rope.cc]: ... by k and v, but got q_num_head is %d, k_num_head is %d

File: paddle/phi/infermeta/spmd_rules/flash_attention.cc [hunk 2]
```diff
@@ -700,6 +718,16 @@ SpmdInfo FlashAttGradInferSpmd(const DistMetaTensor& q,
   auto softmax_lse_dist_attr_dst =
       UnShardTensorDims(softmax_lse_dist_attr, {2});
 
+  if (!is_same_num_heads) {
```
Comment (zhiqiu) [inline paddle/phi/infermeta/spmd_rules/flash_attention.cc]: There are some cases num_head dim can  be sharded.

File: paddle/phi/infermeta/spmd_rules/fused_rope.cc [hunk 3]
```diff
@@ -279,6 +316,10 @@ SpmdInfo FusedRopeInferSpmd(const DistMetaTensor& q,
         UnShardTensorDims(q_dist_attr_dst, {kSeqlenDimIndex, kHeadDimIndex});
   }
 
+  if (!is_same_num_heads) {
+    q_dist_attr_dst = UnShardTensorDims(q_dist_attr_dst, {kNumHeadsDimIndex});
```
Comment (zhiqiu) [inline paddle/phi/infermeta/spmd_rules/fused_rope.cc:339]: same for flash_attn, need to support shard

File: paddle/phi/infermeta/spmd_rules/fused_rope.cc (hunk 1 above)
Comment (deepllz) [reply paddle/phi/infermeta/spmd_rules/fused_rope.cc]: Done

File: paddle/phi/infermeta/spmd_rules/flash_attention.cc (hunk 2 above)
Comment (deepllz) [reply paddle/phi/infermeta/spmd_rules/flash_attention.cc]: 实际测试了一下flash_attention的GQA逻辑，确定是顺序分组，因此修改了一下策略：当num_head能整除切分的维度的是时候，可以支持切分状态。

File: paddle/phi/infermeta/spmd_rules/fused_rope.cc (hunk 3 above)
Comment (deepllz) [reply paddle/phi/infermeta/spmd_rules/fused_rope.cc:339]: 同上

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**